### PR TITLE
[WIP - Do not merge] Alpine Linux Port

### PR DESCRIPTION
**Pull request by benpye**
This currently gets PAL to compile on Alpine Linux, a musl and busybox based distribution. Toward #917 .

I would like some feedback on the method used to set distribution, the build script needs to know as certain libraries for example lttng are unavailable on musl, and whilst coreclr treats them as dependencies, features can be disabled to stop their use. Additionally my securecrt changes are a hack to get compilation working, what would the correct solution to those typedefs be? And I assume I should correct the reason the define is needed, without __GNUC_VA_LIST there are missing symbol errors.

### CODE REVIEW

**Comment by jasonwilliams200OK** [comment]
Thanks @benpye!
@krytarowski, do we know anyone familiar with musl or Alpine in general who can provide some feedback on this work item? Would really help us move forward in this dimension. :)

**Comment by benpye** [comment]
Last I checked @jasonwilliams200OK this was functional on Alpine though likely regressed now. I actually got libunwind packaged for Alpine too. Any ideas for answers to my previous questions would very much be appreciated and I'd happily try and incorporate them.

**Comment by krytarowski** [comment]
I think I recall only one NetBSD user knowing Alpine - @fcambus.

**Comment by jasonwilliams200OK** [comment]
Thanks @krytarowski. Hopefully we get some inputs from @fcambus. Pinging @ncopa from Alpine port project (https://github.com/alpinelinux/aports) to seek help for Ben's question. :)

If someone is wondering, here is another reason to encourage Alpine Linux as Docker is probably switching to Alpine Linux due to its performant and significantly less memory consumption nature http://www.eweek.com/virtualization/slideshows/alpine-linux-goes-all-in-for-docker.html. So running CoreCLR on such a Docker container natively would be considered a great deal!

Here is an open issue for lldb port by @matthewvalimaki: https://bugs.alpinelinux.org/issues/4966.

File: src/pal/src/loader/module.cpp [hunk 1]
```diff
@@ -57,8 +57,10 @@ Module Name:
 #include <sys/types.h>
 #include <sys/mman.h>
 
-#if defined(__LINUX__)
+#if defined(__LINUX__) && defined(__GLIBC__)
 #include <gnu/lib-names.h>
+#else
+#define LIBC_SO "libc.so"
```
Comment (krytarowski) [inline src/pal/src/loader/module.cpp]: Do you really want to define it for NetBSD? https://github.com/dotnet/coreclr/blob/49f39af14e6e2d4048f3c74db0c5164c0f4131b5/src/pal/src/loader/module.cpp#L1625

File: src/pal/src/thread/context.cpp [hunk 2]
```diff
@@ -422,7 +422,7 @@ void CONTEXTToNativeContext(CONST CONTEXT *lpContext, native_context_t *native)
 
     if ((lpContext->ContextFlags & CONTEXT_FLOATING_POINT) == CONTEXT_FLOATING_POINT)
     {
-#ifdef _AMD64_
+#if 0 //_AMD64_
```
Comment (krytarowski) [inline src/pal/src/thread/context.cpp]: why for everybody?

File: src/pal/src/thread/context.cpp [hunk 3]
```diff
@@ -481,7 +481,7 @@ void CONTEXTFromNativeContext(const native_context_t *native, LPCONTEXT lpContex
     
     if ((contextFlags & CONTEXT_FLOATING_POINT) == CONTEXT_FLOATING_POINT)
     {
-#ifdef _AMD64_
+#if 0//def _AMD64_
```
Comment (krytarowski) [inline src/pal/src/thread/context.cpp]: why for everybody?

File: src/pal/src/safecrt/safecrt_output_l.c [hunk 4]
```diff
@@ -21,6 +21,7 @@
 ****/
 
 #define _SAFECRT_IMPL
+#define __GNUC_VA_LIST
```
Comment (krytarowski) [inline src/pal/src/safecrt/safecrt_output_l.c]: Can you define it via CMake? It looks wrong here.

File: src/pal/src/safecrt/output.inl [hunk 5]
```diff
@@ -28,6 +28,7 @@
 //typedef __int64_t __int64;
 
 
+#define __GNUC_VA_LIST
```
Comment (krytarowski) [inline src/pal/src/safecrt/output.inl]: Can you define it via CMake? It looks wrong here.

File: src/pal/src/thread/context.cpp (hunk 3 above)
Comment (benpye) [inline src/pal/src/thread/context.cpp]: This code is generally a quick hack to run on Alpine, as I said. I don't
know a good way to detect musl or Alpine to do this.

File: src/pal/src/safecrt/internal_securecrt.h [hunk 6]
```diff
@@ -24,6 +24,11 @@
 #ifndef _INC_INTERNAL_SECURECRT
 #define _INC_INTERNAL_SECURECRT
 
+typedef unsigned long __uint64_t;
```
Comment (krytarowski) [inline src/pal/src/safecrt/internal_securecrt.h]: This is asking for breakage.

```
$ grep -r __int64_t /usr/include/                                                                                                                     
/usr/include/machine/ansi.h:#define	_BSD_TIME_T_		__int64_t	/* time() */
/usr/include/machine/int_types.h:typedef	long int		__int64_t;
/usr/include/sys/ansi.h:typedef __int64_t	__off_t;	/* file offset */
/usr/include/sys/ansi.h:	__int64_t __mbstateL; /* for alignment */
/usr/include/sys/socket.h:#define _SS_ALIGNSIZE	(sizeof(__int64_t))
/usr/include/sys/socket.h:	__int64_t     __ss_align;/* force desired structure storage alignment */
/usr/include/sys/stdint.h:typedef	__int64_t	int64_t;
/usr/include/sys/stdint.h:#define	int64_t		__int64_t
/usr/include/sys/types.h:typedef	__int64_t	int64_t;
/usr/include/sys/types.h:#define	int64_t		__int64_t
/usr/include/sys/common_ansi.h:#define	_BSD_TIME_T_		__int64_t	/* time() */
/usr/include/sys/common_int_types.h:typedef	__INT64_TYPE__		  __int64_t;
/usr/include/stdint.h:typedef	__int64_t	int64_t;
/usr/include/stdint.h:#define	int64_t		__int64_t
/usr/include/amd64/ansi.h:#define	_BSD_TIME_T_		__int64_t	/* time() */
/usr/include/amd64/int_types.h:typedef	long int		__int64_t;
/usr/include/i386/ansi.h:#define	_BSD_TIME_T_		__int64_t	/* time() */
/usr/include/i386/int_types.h:typedef	__COMPILER_INT64__	__int64_t;
/usr/include/i386/int_types.h:typedef	long long int		__int64_t;
```

File: src/pal/CMakeLists.txt [hunk 7]
```diff
@@ -2,6 +2,8 @@ cmake_minimum_required(VERSION 2.8.12.2)
 
 project(COREPAL)
 
+set(PAL_CMAKE_ENABLE_TRACE 0)
```
Comment (krytarowski) [inline src/pal/CMakeLists.txt]: Are you disabling it for everybody?

File: src/ToolBox/SOS/lldbplugin/CMakeLists.txt [hunk 8]
```diff
@@ -17,6 +17,11 @@ elseif(CLR_CMAKE_PLATFORM_ARCH_ARM64)
     return()
 endif()
 
+if(CLR_CMAKE_PLATFORM_LINUX_ALPINE)
+    message(WARNING "lldb is not supported on Alpine Linux yet")
```
Comment (krytarowski) [inline src/ToolBox/SOS/lldbplugin/CMakeLists.txt]: I don't understand conditions against few of few hundreds of Linux distributions...

File: CMakeLists.txt [hunk 9]
```diff
@@ -705,4 +714,4 @@ else ()
   endif ()
 endif (IS_64BIT_BUILD EQUAL 1)
 
-add_subdirectory(src)
+#add_subdirectory(src)
```
Comment (krytarowski) [inline CMakeLists.txt]: Why? It doesn't look correct.

**Comment by krytarowski** [comment]
In general I don't like this code. Can you use a saner detectable libc library on Alpine? Perhaps go for rump unikernel and use NetBSD library stack on Alpine it will be way cleaner.

**Comment by ncopa** [comment]
> In general I don't like this code.

+1

However, it gives a hint on what needs fixing, but I don't think this should be merged as-is.

In general, it is better to test for features than test for platforms, and then in the code do something like:

````C
#if HAVE_FEATURE_THIS
...
#elif HAVE_FEATURE_THAT
...
#else
/* fallback to some standard, like POSIX */
...
#endif
````

**Comment by jasonwilliams200OK** [comment]
Thanks for the comments guys. I am configuring Alpine Linux 3.3.1 at the moment. When i will get clang and dependencies in place, will try to feature detect to replace those hardcoded ifdef's in CMake, the way it is done at other places in coreclr. Stay tuned! :)

@ncopa, is anyone working on porting lldb to Alpine at the moment? @krytarowski, ported lldb recently to NetBSD via pkgsrc, so we can probably get some collaboration in this area. :)

**Comment by ellismg** [comment]
@jasonwilliams200OK You may be able to unblock yourself by making the LLDB requirement optional when LLDB is not installed.  We do this on OSX.  We won't build the SOS plugin which is not great long term because it's helpful in debugging the runtime but it's not a blocker for getting the runtime to boot.

**Comment by jasonwilliams200OK** [comment]
Thanks @ellismg. I will use Ben's code from this PR where ldd is already disabled for Alpine.
While I was doing patches for CoreFX native, I started to work on Alpine and able to build 50% so far. Some networking stack modules use different headers and feature detection is bit tricky as while Alpine is categorized as Linux by CMake, most of the lowlevel headers are different (especially `linux/*.h` ones).

**Comment by jasonwilliams200OK** [comment]
@ncopa, I am able to build 97% of CoreFX (the .NET Core base libraries repo which has comparatively a small native code base) on Alpine with this patch https://github.com/jasonwilliams200OK/corefx/commit/63e61a36e9585255844f6c141b39d4214bd53634. Once I get that working, I will try to contribute to CoreCLR (which is the actual runtime).

To perform a quick compile:

```sh
# in alpine ash, bash or sh
git clone https://github.com/jasonwilliams200ok/corefx -b alpine-linux-port --single-branch
cd corefx
./build.sh native  # requires cmake, libopenssl and couple of other dependencies
                   # which i found using `apk search`
```

The remaining portion is where the networking layer of CoreFX relies on netlink headers, for example nlmsghdr is used from netlink header.h, which are missing from . At this point, can I get some help to figure out the remaining issues? Or better yet can we add CoreFX to aports and fix the remaining build issues there and then upstream the patches?

**Comment by jasonwilliams200OK** [comment]
With https://github.com/jasonwilliams200OK/corefx/commit/337f5ed4927f210a6b1b5bbda70792da53d7cd73 patch, CoreFX' System.Native now builds fully on Alpine Linux.

Thanks to @ncopa for pointing me out that we need `apk add linux-headers` to get netlink headers (chatted through IRC freenode#alpine). 

Thanks to @gyrosgeier for pointing me out that there is a better (non-cmake) way to detect type disparities across OSes implementations of same struct member on freenode#cmake (this partial patch https://github.com/jasonwilliams200OK/corefx/commit/939fc312474afee11707e7759f6a491587b40510).

I will also try to refactor the type disparities across CoreFX as a general purpose `TypeProvider` with all the required `decltype`s and a template-based `GetResolvedTypeValue(T)`, as @nguerrera once envisioned. :)

Now it's time to up-streaming smaller patches to CoreFX repo for musl-libc support! :zap:

**Comment by nguerrera** [comment]
> I will also try to refactor the type disparities across CoreFX as a general purpose TypeProvider with all the required decltypes and a template-based GetResolvedTypeValue(T), as @nguerrera once envisioned. :)

I'm actually not sure what you're referring to here. Neither TypeProvider nor GetResolvedTypeValue ring any bells...

**Comment by jasonwilliams200OK** [comment]
@nguerrera, actually those names are my own, I was referring to this https://github.com/dotnet/corefx/pull/3479#issuecomment-143504046. Sorry for the confusion. :smile:

File: src/pal/src/loader/module.cpp (hunk 1 above)
Comment (benpye) [inline src/pal/src/loader/module.cpp]: This was a quick change to get it working, not sure of the correct path forward.

File: src/pal/src/thread/context.cpp (hunk 2 above)
Comment (benpye) [inline src/pal/src/thread/context.cpp]: Need a suitable CMake test.

File: src/pal/CMakeLists.txt (hunk 7 above)
Comment (benpye) [inline src/pal/CMakeLists.txt]: Don't have a good way to toggle this.

File: src/ToolBox/SOS/lldbplugin/CMakeLists.txt (hunk 8 above)
Comment (benpye) [inline src/ToolBox/SOS/lldbplugin/CMakeLists.txt]: Not sure the best way forward, guess either just make it optional everywhere or have a nolldb parameter to build.sh . Just wanted to move forward. `CLR_CMAKE_PLATFORM_LINUX_ALPINE` should probably be a musl toggle.

File: CMakeLists.txt (hunk 9 above)
Comment (benpye) [inline CMakeLists.txt]: Didn't intend to commit that, reducing my own build time while getting PAL working.

**Comment by benpye** [comment]
Added some comments to address some issues. As you may of noticed this was sort of a dump of my current progress, unfortunately I haven't moved forward with it since. Most of the issues above are things we need to either test for, or toggle at build time some how.

**Comment by jasonwilliams200OK** [comment]
@benpye, @jkotas, this can be closed now, as all the patches for Alpine Linux are upstreamed and CoreCLR fully builds (modulo lldbplugin): #917

**Comment by jkotas** [comment]
@jasonwilliams200OK @benpye  Thanks a lot to plowing through all the Alpine issues!

**Pull request closed without merging**